crates/ra_ide/src/completion/completion_context.rs

```diff
@@ -95,15 +104,45 @@ impl<'a> CompletionContext<'a> {
             has_type_args: false,
             dot_receiver_is_ambiguous_float_literal: false,
         };
-        ctx.fill(&original_file, file_with_fake_ident, position.offset);
+
+        let mut original_file = original_file.syntax().clone();
+        let mut hypothetical_file = file_with_fake_ident.syntax().clone();
+        let mut offset = position.offset;
+        let mut fake_ident_token = fake_ident_token;
+
+        // Are we inside a macro call?
+        while let (Some(actual_macro_call), Some(macro_call_with_fake_ident)) = (
```

Comment:
Originally I've thought that recusively expanding the macros is exactly how we should do IDE features. However, @edwin0cheng had an idea which seems like it kinda can be better? In syntax highlighting, we descend to the bottom of macro expansion in one step:

https://github.com/rust-analyzer/rust-analyzer/blob/aff82cf7ac172f213cb5dcca637cb2c5332294c1/crates/ra_ide/src/syntax_highlighting.rs#L109

I think *ideally* this is what we want to do for completion as well. Although I am not sure if `expand_hypothetical` can be made to work this way, especially if we switch from "resolve" to "record" strategy for SourceAnalyzer

Reply:
Some vague thoughs for *negative* durability in salsa for such hypothetical cases...

Reply:
Support in salsa for throwaway 'alternative universes' would be cool...

Reply:
But about the recursive expanding, I wanted to avoid a situation where the 'hypothetical' expansion is too different; I think we need to 'follow along' both to keep track. There are probably a lot of ways this can currently break if the inserted token changes the expansion too much.